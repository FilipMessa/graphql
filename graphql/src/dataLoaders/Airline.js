// @flow

import airlines from '../types/Airlines.json';
import AirlineType from '../types/Airline';

export function createAirline(airlineCode: string): AirlineType {
  const airline = airlines.find(airline => airline.id === airlineCode);
  return {
    name: airline.name,
    code: airlineCode,
    logoUrl: `https://images.kiwi.com/airlines/64/${airlineCode}.png`,
    isLowCost: airline.lcc === '1',
  };
}
