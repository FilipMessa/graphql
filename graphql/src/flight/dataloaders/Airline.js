// @flow

import Dataloader from 'dataloader';
import { get } from '../../services/HttpRequest';
import Config from '../../../config/application';

import type { Airline as AirlineType } from '../Flight';

let airlinesCache = null;

type Airline = {
  name: string,
  id: string,
  lcc: string,
};

export default function createAirlineLoader() {
  airlinesCache = null;
  return new Dataloader((airlinesCodes: Array<string>) =>
    batchLoad(airlinesCodes),
  );
}

async function batchLoad(airlinesCodes): Promise<Array<?AirlineType | Error>> {
  const airlines = await fetchAirlines();
  return airlinesCodes.map(airlineCode => {
    const airline = airlines[airlineCode];
    return airline !== undefined ? createAirline(airline) : null;
  });
}

async function fetchAirlines() {
  if (!airlinesCache) {
    airlinesCache = get(Config.restApiEndpoint.airlines)
      // eslint-disable-next-line promise/prefer-await-to-then
      .then(airlines => {
        const cache: { [key: string]: Airline } = {};
        airlines.map((a: Airline) => (cache[a.id] = a));
        return cache;
      });
  }
  return airlinesCache;
}

function createAirline(airline: Airline): AirlineType {
  return {
    name: airline.name,
    code: airline.id,
    logoUrl: `https://images.kiwi.com/airlines/64/${airline.id}.png`,
    isLowCost: airline.lcc === '1',
  };
}
