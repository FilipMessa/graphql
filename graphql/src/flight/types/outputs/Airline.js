// @flow

import { GraphQLObjectType, GraphQLString, GraphQLBoolean } from 'graphql';

import type { Airline } from '../../Flight';

export default new GraphQLObjectType({
  name: 'Airline',
  fields: {
    name: {
      type: GraphQLString,
      resolve: ({ name }: Airline): string => name,
    },

    code: {
      type: GraphQLString,
      description: '2-letter IATA code of airline.',
      resolve: ({ code }: Airline): string => code,
    },

    logoUrl: {
      type: GraphQLString,
      resolve: ({ logoUrl }: Airline): string => logoUrl,
    },

    isLowCost: {
      type: GraphQLBoolean,
      description: 'Airline is low cost carrier.',
      resolve: ({ isLowCost }: Airline): boolean => isLowCost,
    },
  },
});
