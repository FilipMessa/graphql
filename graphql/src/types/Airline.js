// @flow

import { GraphQLObjectType, GraphQLString, GraphQLBoolean } from 'graphql';

import type { AirlineType } from '../Entities';

export default new GraphQLObjectType({
  name: 'Airline',
  fields: {
    name: {
      type: GraphQLString,
      resolve: ({ name }: AirlineType): string => name,
    },

    code: {
      type: GraphQLString,
      description: '2-letter IATA code of airline.',
      resolve: ({ code }: AirlineType): string => code,
    },

    logoUrl: {
      type: GraphQLString,
      resolve: ({ logoUrl }: AirlineType): string => logoUrl,
    },

    isLowCost: {
      type: GraphQLBoolean,
      description: 'Airline is low cost carrier.',
      resolve: ({ isLowCost }: AirlineType): boolean => isLowCost,
    },
  },
});
