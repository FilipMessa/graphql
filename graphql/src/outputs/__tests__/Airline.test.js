// @flow

import Airline from '../../flight/types/outputs/Airline';

it('Airline type should have valid fields', () => {
  expect(Airline.getFields()).toMatchSnapshot();
});
