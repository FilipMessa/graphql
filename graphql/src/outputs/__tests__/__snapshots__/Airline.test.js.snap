// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Airline type should have valid fields 1`] = `
Object {
  "code": Object {
    "args": Array [],
    "description": "2-letter IATA code of airline.",
    "isDeprecated": false,
    "name": "code",
    "resolve": [Function],
    "type": "String",
  },
  "isLowCost": Object {
    "args": Array [],
    "description": "Airline is low cost carrier.",
    "isDeprecated": false,
    "name": "isLowCost",
    "resolve": [Function],
    "type": "Boolean",
  },
  "logoUrl": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "logoUrl",
    "resolve": [Function],
    "type": "String",
  },
  "name": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "name",
    "resolve": [Function],
    "type": "String",
  },
}
`;
